cmd/util/ledger/reporters/resource_reporter.go

```diff
@@ -0,0 +1,181 @@
+	rwf      ReportWriterFactory
+	rw       ReportWriter
+	progress *progressbar.ProgressBar
+}
+
+func NewConstantSizeReporter(logger zerolog.Logger, rwf ReportWriterFactory) *ConstantSizeReporter {
+	return &ConstantSizeReporter{
+		log: logger,
+		rwf: rwf,
+	}
+}
+
+func (r *ConstantSizeReporter) Name() string {
+	return "Resource Reporter"
+}
+
+func (r *ConstantSizeReporter) Report(payloads []ledger.Payload, commit ledger.State) error {
+	r.rw = r.rwf.ReportWriter(ConstantSizeReporterReportPrefix)
+	defer r.rw.Close()
+
+	wg := &sync.WaitGroup{}
+
+	// we need to shard by owner, otherwise ledger won't be thread-safe
+	addressCount := 0
+	payloadsByOwner := make(map[flow.Address][]ledger.Payload)
+
+	for _, pay := range payloads {
+		owner := flow.BytesToAddress(pay.Key.KeyParts[0].Value)
+		if len(owner) > 0 { // ignoring payloads without ownership (fvm ones)
+			m, ok := payloadsByOwner[owner]
+			if !ok {
+				payloadsByOwner[owner] = make([]ledger.Payload, 0)
+				addressCount++
+			}
+			payloadsByOwner[owner] = append(m, pay)
+		}
+	}
+
+	jobs := make(chan job, addressCount)
+	r.progress = progressbar.Default(int64(addressCount), "Processing:")
+
+	for k, v := range payloadsByOwner {
+		jobs <- job{k, v}
+	}
+
+	close(jobs)
+	workerCount := runtime.NumCPU()
+	for i := 0; i < workerCount; i++ {
+		wg.Add(1)
+		go r.worker(jobs, wg)
+	}
+
+	wg.Wait()
+
+	r.log.Info().Msgf(">>>>>> %d of %d", numberOfCompositeTypesWithConstantSizeChildren, numberOfCompositeTypes)
+
+	err := r.progress.Finish()
+	if err != nil {
+		panic(err)
+	}
+
+	return nil
+}
+
+func (r *ConstantSizeReporter) worker(
+	jobs <-chan job,
+	wg *sync.WaitGroup) {
+	for j := range jobs {
+
+		view := migrations.NewView(j.payloads)
+		st := state.NewState(view, state.WithMaxInteractionSizeAllowed(math.MaxUint64))
+		sth := state.NewStateHolder(st)
+		accounts := state.NewAccounts(sth)
+		storage := cadenceRuntime.NewStorage(
+			&migrations.AccountsAtreeLedger{Accounts: accounts},
+			nil,
+		)
+
+		owner, err := common.BytesToAddress(j.owner[:])
+		if err != nil {
+			panic(err)
+		}
+
+		inter := &interpreter.Interpreter{}
+		for _, domain := range domains {
+			storageMap := storage.GetStorageMap(owner, domain, true)
+			itr := storageMap.Iterator(inter)
+			_, value := itr.Next()
+			for value != nil {
+				interpreter.InspectValue(
+					inter,
+					value,
+					func(v interpreter.Value) bool {
+						if composite, ok := v.(*interpreter.CompositeValue); ok {
+							atomic.AddUint64(&numberOfCompositeTypes, 1)
+							if allFieldsAreConstantSize(r.log, inter, composite) {
+								atomic.AddUint64(&numberOfCompositeTypesWithConstantSizeChildren, 1)
+							}
+						}
+
+						return true
+					},
+				)
+
+				_, value = itr.Next()
+			}
+		}
+
+		err = r.progress.Add(1)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	wg.Done()
+}
+
+func allFieldsAreConstantSize(logger zerolog.Logger, inter *interpreter.Interpreter, value *interpreter.CompositeValue) bool {
```

Comment:
I think this recursive check might have to be done on the 'type' (i.e: `CompositeType`) rather than the 'value' (`CompositeValue`).
The reason is, there could be optional fields (`T?`), and the stored value is `nil`. Even though the `nil` value is not expandable, the potential value space for that field is expandable.
e.g:
```cadence
struct Foo {
    var values: [Bar]?
}
```
Here if the value was constructed as `Foo {values: nil}` the reporter would return as if it is inline-able. But in reality, it may be not.

Reply:
How do I get to the type using the iterator ?